content/blog/managing-google-workspace-with-terraform.md

```diff
@@ -0,0 +1,295 @@
+            }
+          }
+        }
+      }
+      groups = {
+        "test-group" = {
+          name  = "Test Group"
+          email = "test-group@example.com"
+        }
+      }
+    }
+    expect_failures = [var.users]
+  }
+  ```
+
+### Design Decision #2 - Choosing Intuitive Terraform Variable Structure
+
+In the Google Workspace provider, provisioning a group involves declaring two resources: `group` and `group_settings`.
+
+```terraform
+resource "googleworkspace_group" "sales" {
+  email = "sales@example.com"
+}
+
+resource "googleworkspace_group_settings" "sales_settings" {
+  email        = googleworkspace_group.sales.email
+  who_can_join = "INVITED_CAN_JOIN"
+}
+```
+
+This `group` and `group_setting` resource design mirrors Google's Admin SDK REST API structure. However, we felt that group settings more intuitively belonged as a nested attribute inside a group. So in our group variable, we added `settings` and extracted settings in the module's `local` block:
+
+```terraform
+# groups variable declaration
+variable "groups" {
+  description = "List of groups"
+  type = map(object({
+    name: string,
+    email: string,
+    settings: optional(object({
+      who_can_join: optional(string),
+    }))
+  }))
+}
+
+# locals block
+locals {
+  group_settings = {
+    for k, v in var.groups : k => merge(v.settings, { email = v.email })
+  }
+}
+
+# group_settings resource
+resource "googleworkspace_group_settings" "defaults" {
+  for_each     = local.group_settings
+  email        = each.value.email
+  who_can_join = "INVITED_CAN_JOIN"
+}
+```
+
+Yes, it's an abstraction which can sometimes lead to issues, but we think reducing cognitive friction is worth the added business logic in the Terraform code. Here's an example of a simpler configuration for groups:
+
+```terraform
+locals {
+  default_group_settings = {
+    who_can_join = "ALL_IN_DOMAIN_CAN_JOIN"
+  }
+}
+
+module "googleworkspace" {
+  source  = "masterpointio/users-groups-automation/googleworkspace"
+  version = "X.X.X"
+
+  groups = {
+    support = {
+      name    = "Support"
+      email   = "support@example.com"
+      settings = merge(local.default_group_settings, {})
+    }
+    engineers = {
+      name    = "Engineering"
+      email   = "engineering@example.com"
+      settings = merge(local.default_group_settings, {
+        who_can_join = "INVITED_CAN_JOIN"
+      })
+    }
+  }
+}
+```
+
+## Wrapping Up
+
+With this setup in place and a CODEOWNERS file, we're now transparently, democratically, and securely managing users and groups within Google Workspace. This helps our engineering leader, Matt, avoid having to keep this in his head and enables us as a team to avoid waiting on him. Want access to AWS SSO? Open a PR and get the right approval from the CODEOWNERS, merge it, and automation takes care of the rest. We love this workflow and we hope it's useful to you and your team!
+
+Want help implementing this terraform module set in your Google Workspace?
```

Comment:
```suggestion
Want help automating your Google Workspace configuration with Terraform? Need to make sure your Workspace policies stay consistent and auditable? Want to avoid onboarding chokepoints like Matt used to be?
```

Or similar?